codesearch/server/server_test.go

```diff
@@ -0,0 +1,128 @@
+	"github.com/stretchr/testify/require"
+
+	gitpb "github.com/buildbuddy-io/buildbuddy/proto/git"
+	inpb "github.com/buildbuddy-io/buildbuddy/proto/index"
+	spb "github.com/buildbuddy-io/buildbuddy/proto/search"
+)
+
+func mustMakeServer(t *testing.T) *codesearchServer {
+	tmpDir := testfs.MakeTempDir(t)
+
+	te := real_environment.NewRealEnv(nil)
+	te.SetAuthenticator(&nullauth.NullAuthenticator{})
+
+	server, err := New(te, tmpDir, tmpDir)
+	require.NoError(t, err)
+	return server
+}
+
+func bootstrapIndex(t *testing.T, ctx context.Context, server *codesearchServer, repoURL, sha string) {
+	// In order to apply incremental updates, we need to first set the last indexed commit
+	ns, err := server.getUserNamespace(ctx, "")
+	require.NoError(t, err)
+	iw, err := index.NewWriter(server.db, ns)
+	require.NoError(t, err)
+	ru, err := git.ParseGitHubRepoURL(repoURL)
+	require.NoError(t, err)
+	github.SetLastIndexedCommitSha(iw, ru, sha)
+	err = iw.Flush()
+	require.NoError(t, err)
+}
+
+func TestIncrementalIndex(t *testing.T) {
+	ctx := context.Background()
+	server := mustMakeServer(t)
+
+	oldSha := "def456"
+	newSha := "abc123"
+	bootstrapIndex(t, ctx, server, "github.com/buildbuddy-io/buildbuddy", oldSha)
+
+	rsp, err := server.Index(ctx, &inpb.IndexRequest{
+		GitRepo: &gitpb.GitRepo{
+			RepoUrl: "github.com/buildbuddy-io/buildbuddy",
+		},
+		ReplacementStrategy: inpb.ReplacementStrategy_INCREMENTAL,
+		Update: &inpb.IncrementalUpdate{
+			Commits: []*inpb.Commit{
+				{
+					Sha:       newSha,
+					ParentSha: oldSha,
+					AddsAndUpdates: []*inpb.File{
+						{
+							Filepath: "foo/bar/baz.txt",
+							Content:  []byte("doo be doo be doooo"),
+						},
+					},
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.NotNil(t, rsp)
+
+	repoStatus, err := server.RepoStatus(context.Background(), &inpb.RepoStatusRequest{
+		RepoUrl: "github.com/buildbuddy-io/buildbuddy",
+	})
+	require.NoError(t, err)
+	assert.Equal(t, repoStatus, &inpb.RepoStatusResponse{
+		LastIndexedCommitSha: newSha,
+	})
+
+	searchRsp, err := server.Search(context.Background(), &spb.SearchRequest{
+		Query: &spb.Query{
+			Term: "doo",
+		},
+	})
+	require.NoError(t, err)
+
+	assert.NotNil(t, searchRsp)
+	assert.NotNil(t, searchRsp.Results)
+	assert.Len(t, searchRsp.Results, 1)
+	search := searchRsp.Results[0]
+	search.Snippets = nil // clear out snippets to allow for easier comparison
+	assert.Equal(t, search, &spb.Result{
+		Owner:      "buildbuddy-io",
+		Repo:       "buildbuddy",
+		Sha:        newSha,
+		Filename:   "foo/bar/baz.txt",
+		MatchCount: 1,
+	})
+}
+
+func TestIndexFull(t *testing.T) {
+	t.Skip("This operation downloads a github repo, and requires refactoring to unit test.")
+}
+
+func TestRepoStatus_NoStatus(t *testing.T) {
+	server := mustMakeServer(t)
+
+	response, err := server.RepoStatus(context.Background(), &inpb.RepoStatusRequest{
+		RepoUrl: "github.com/buildbuddy-io/buildbuddy",
+	})
+	assert.True(t, status.IsNotFoundError(err))
```

Comment:
Could this use `assert.ErrorIs`? Otherwise assertion failures would have pretty bad messages.

Reply:
Oh, I didn't know about ErrorIs, thanks. An unfortunate aspect of it, though, is that it requires you to construct an equivalent error, including the specific error message, so the test becomes more brittle. I've updated this to add a better message to the assert (which seems to be the more common pattern in the BB codebase for status errors- ErrorIs / `errors.Is` are used rarely).